File: Tree/BST/[230]kthSmallestElementInABST.java
class Solution {
  Deque<TreeNode> stack = new LinkedList<>();
  public int kthSmallest(TreeNode root, int k) {
    TreeNode node = root;
    while (true) {
      while (node != null) {
        stack.addLast(node);
        node = node.left;
      }
      node = stack.removeLast();
      if (--k == 0) return node.val;
      node = node.right;
    }
  }
}

/**
 * performance: 0 ms < 100%
 * time complexity: O(n)
 * space complexity: O(depth)
 * notes: in-order traversal using stack.
 */


class Solution {
  Deque<TreeNode> stack = new LinkedList<>();
  public int kthSmallest(TreeNode root, int k) {
    TreeNode next = root;
    TreeNode node = null;
    for (int i = 0; i < k; i++) {
      helper(next);
      node = stack.removeLast();
      next = node.right;
    }
    return node.val;
  }
  private void helper(TreeNode node) {
    while (node != null) {
      stack.addLast(node);
      node = node.left;
    }
  }
}

/**
 * same
 */

File: Tree/BST/[230]kthSmallestElementInABST.py
class Solution:
  def kthSmallest(self, root: TreeNode, k: int) -> int:
    res = 0
    def helper(node):
        nonlocal k, res
        if node.left: helper(node.left)
        if k == 0: return
        k -= 1
        if k == 0:
            res = node.val
            return
        if node.right: helper(node.right)
    helper(root)
    return res

# performance: 48 ms < 99%
# time complexity: O(height + k)
# space complexity: O(height + k)
